File: language/scala/src/main/resources/scala-code/tree/Leaf.scala
package tree

class Leaf(val value:Any) extends Tree {

  override def canEqual(a: Any) = a.isInstanceOf[Leaf]

  override def asLeaf() : Option[Leaf] = Some(this)

  /** link in with default equals. */
  override def equals(tree: Any): Boolean = {
    tree match {
      case o: Leaf => o.canEqual(this) && same(o)

      case _ => false
    }
  }
}

File: src/main/resources/java-code/Leaf.java
public class Leaf implements Tree {
    public final Object value;

    public boolean equals (Object o) {
        if (o == null) { return false; }
        if (o instanceof Tree) {
            return same((Tree) o);
        }
        return false;
    }

    public Leaf(Object e) {
        value = e;
    }

    public java.util.Optional<Leaf> asLeaf() { return java.util.Optional.of(this); }

    public String toString() {
        return "(Leaf:" + value + ")";
    }
}
